### PR TITLE
Written an article titled "Symfony quickstart"

### PR DESCRIPTION
**Pull request by lajosarpad**

### CODE REVIEW

File: 2020/02/24/symfony-quickstart.html.md [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+
+Using template engines instead of allowing the developers to use PHP as a template engine has the drawback of limiting the options, however, the exact reason they are used, especially by large teams is the exact same thing that we can outline as its drawback. Yes, it limits the options of the developers, but this way developers will not find it so easy and convenient to temporarily (?) implement their business logic as part of the view.
+
+If my opinion matters, I’m okay with using PHP as a template engine as long as we do not mix up the layers of the application and, by convention the view will remain “only” the view. I’m also okay with using template engines, even though it is less convenient to use when we need to do some experimenting and writing some code for a very short time, after which it is reverted. In that kind of experimental work one feels that enforcing this separation is unnecessary bureaucracy, but the gain is that we will not have to deal with views where lots of things are implemented which should not be there. So, this is a harmless form of bureaucracy, if such a thing exists.
```
Comment (jonjensen) [inline 2020/02/24/symfony-quickstart.html.md]: Remove "If my opinion matters," -- this is your blog post, and your opinion matters!

I think you nicely weighed the pros and cons of PHP vs. other templating languages and your opinion makes perfect sense here.

File: 2020/02/24/symfony-quickstart.html.md [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+
+## Summary
+
+Symfony is a Framework which provides us with a way to work in conformity with the MVC pattern. Models are Entity classes along with their Repository pair, representing database tables, while Doctrine watches over the whole process, unless one chooses not to use Doctrine. Views are templates that generate HTML, for which Twig is a convenient tool, but one can use Symfony without relying on the usage of Twig. Configuration can be written in YAML, XML or PHP. PHP dependencies are handled by Composer, client-side dependencies are handled by Yarn. Symfony has its own console that can be used to run different commands.
```
Comment (jonjensen) [inline 2020/02/24/symfony-quickstart.html.md]: Again, client-side dependencies may be nonexistent, handled by Yarn, or by other things. AFAIK Symfony doesn't have any stance on that, right?

File: 2020/02/24/symfony-quickstart.html.md [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+
+```
+yarn run encore <env> --watch
+```
+
+and from there on, changes will be watched and detected, so one can be confident that client-side code will end up in the target path and will not have to be worried about it.
+
+The drawback, of course, is that this will eat up resources and will still be time-consuming if the client-side code is very large. As a result, it is a much smarter way to organize code to just use the source path while in development mode. Sometimes browser cache will still cause us trouble while working this way, but a seasoned frontend developer is already used to that.
+
+Webpack is a nice tool and it can be used in a very wise manner but it is a good idea to limit its use wisely. Avoiding its usage in development mode could be very helpful, allowing us to nicely debug and not worry about building client-side code. In prod mode it adds a lot of value in obfuscating, minimising and packing the code.
```
[the remark this answers is not in the document]
Comment (lajosarpad) [reply 2020/02/24/symfony-quickstart.html.md]: Indeed, one may choose not to use these frontend tools, but the same can be said about Doctrine, Twig, etc. Symfony allows us to use Doctrine or Twig on the server-side and the same can be said about the above-mentioned client-side tools. It is true that API-only projects don't require any Javascript tools, but, in an analogous manner we can say that no-db Symfony projects don't require Doctrine usage.

https://symfony.com/doc/current/frontend/encore/simple-example.html

So, in short, one may or may not use these components in Symfony, but during my work on Symfony projects I had the impression that there is a preference in using these, which can be felt from the documentation and forum discussions. So, I think these tools deserve a presentation here and readers may decide whether they need them. However, I am happy to add some remarks which suggest that it is not necessary to use these client-side tools.

File: 2020/02/24/symfony-quickstart.html.md [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+
+## MVC
+
+<div style="width: 100%; text-align: center;"><img src="/blog/2020/02/21/symfony-quickstart/mvc.png"></div>
+
+Image taken from [tutorialspoint](https://www.tutorialspoint.com/mvc_framework/mvc_framework_introduction.htm).
```
[the remark this answers is not in the document]
Comment (lajosarpad) [reply 2020/02/24/symfony-quickstart.html.md]: @jonjensen @obnoxiousorc it would be great!

File: 2020/02/24/symfony-quickstart.html.md [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+
+On the other hand, the complex tool that we add is difficult to understand unless one delves into the actual code which generates the schema and the way the schema is generated is out of our hands, unless we are content with the current version of doctrine or accept the price of hacking into the code each time a new version is downloaded.
```
[the remark this answers is not in the document]
Comment (lajosarpad) [reply 2020/02/24/symfony-quickstart.html.md]: That's human sloppyness, which is always an issue. However, if we have two schema versions, then the difference can always be generated. Liquibase is such a tool. So, at deploy time the rollback script could be automatically generated. My opinion is that it's much better to write schema-changing scripts either manually or in a manner that we fully understand what happens than trusting a tool to automatically do it for us. As about the rollback script, which might become urgent, as you have described, I believe it should be automatically generated. That way we do not overcomplicate the way we develop our applications and keep the schema change in our hands as well.

File: 2020/02/24/symfony-quickstart.html.md [hunk 6]
```diff
@@ -0,0 +1,429 @@
...
+
+In short, my advice is to try to avoid generating schema changes based on entity classes whenever possible. We can always find exceptional cases, but in general, SQL is clearer than a tool that generates SQL.
```
[the remark this answers is not in the document]
Comment (lajosarpad) [reply 2020/02/24/symfony-quickstart.html.md]: @jonjensen I wanted to avoid it, so I have intended to emphasize the domain of the argument to make sure it will not be misunderstood. We of course need an ORM in many cases because that level of abstraction on data-level is saving a lot of time. However, in my opinion the line we should not cross is the schema scripting, because the problems we might face are much worse than the problems you have described earlier in the scenario when there is no rollback script in a case of urgency.

If, for example, for any reason the rollback script generated by Doctrine is not working, then we are not only in a situation of urgently resolve the issue with a live db problem, but we do not even know what we want to do.

So, I believe a nice compromise is to approach the issue in the following way:

- we keep schema scripting in our hands, maybe using some tools, but definitely in a manner that we fully understand what happens
- we include a step to generate rollback scripts at deploy time
- we use ORM for data management in most cases
- when the ORM is not smart-enough, we write some scripts on our own, but very carefully and keeping track where such custom scripts are in the source-code

File: 2020/02/24/symfony-quickstart.html.md [hunk 7]
```diff
@@ -0,0 +1,429 @@
...
+
+## Caching
+
+Symfony supports caching with the caching module, which, if switched on will cache configuration and routes. If controller actions or routes are changed, then it is advisable to clear the cache via
+
+```
+bin/console cache:clear && bin/console cache:warmup
+```
+
+Whether or not to switch on caching for your application at dev mode depends on the frequency of a need to cache cleanup. Read more [here].
```
[the remark this answers is not in the document]
Comment (lajosarpad) [reply 2020/02/24/symfony-quickstart.html.md]: Thanks for pointing this out.

File: 2020/02/24/symfony-quickstart.html.md (hunk 2 above)
Comment (lajosarpad) [reply 2020/02/24/symfony-quickstart.html.md]: @jonjensen I don't know about an official stance Symfony would have on the client-side dependencies, but the official documentation describes their support and overall my experience with Symfony is that these specific tools are strongly preferred by the community. Just like Doctrine and Twig is optional, client-side usage of Yarn, for example is optional as well. I will do a change to make sure this cannot be misunderstood.

**Comment by obnoxiousorc** [comment]
@lajosarpad this is now live!
